### ISSUE TITLE
Gradient Penalty implementation in WGANGP

### ISSUE THREAD

**Issue by hello-fri-end** [mlpack/mlpack#2891]
File : [https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/gan/wgangp_impl.hpp#L79](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/gan/wgangp_impl.hpp#L79)

WGAN-GP Loss function of Critic : 
L = D(x) - D(G(z)) + lamda *  ( norm(gradient(x')) -1) ^2

where x'  = epsilon * ( real_image) + (1 - epsilon) * fake_image

So, the critic's loss function requires us to calculate gradient of a gradient in the backward pass, but in the current implementation, although, the term is added to the critic's loss function but it's gradient is not computed. 

Please let me now if I've got it wrong.

**Comment by zoq** [comment]
Interesting catch, looked at the implementation and you are right, we are missing the gradient step.

**Comment by hello-fri-end** [comment]
Also, gradient needs to calculated for epsilon as well. In PyTorch you'd do something like torch.rand(len(real), 1, 1, 1, device=device, requires_grad=True)). I'm not sure how this would be implemented in mlpack..

**Comment by hello-fri-end** [comment]
Hey @zoq, can you please take a look at this? I'm unable to figure this out.

**Comment by zoq** [comment]
Thanks for bringing this up again. It's now on my list for the weekend.

**Comment by zoq** [comment]
@hello-fri-end in the last meeting you mentioned you have to calculate the gradient of the gradient for the whole network, but it's it just the gradient of the gradient of the loss function? You mentioned you asked on stackoverflow as well, so maybe a link might be helpful as well, not sure.

**Comment by hello-fri-end** [comment]
> @hello-fri-end in the last meeting you mentioned you have to calculate the gradient of the gradient for the whole network, but it's it just the gradient of the gradient of the loss function? You mentioned you asked on stackoverflow as well, so maybe a link might be helpful as well, not sure.

 @zoq I did not receive any response on my previous stackoverflow post. I wrote another one today with the code for gradient penalty in Pytorch: (https://stackoverflow.com/questions/67176984/backward-pass-of-gradient-penalty-loss-in-wgangp). Maybe the code will be helpful.

**Comment by hello-fri-end** [comment]
Here's my hand-computed gradient for the loss function: [https://drive.google.com/file/d/1GOD5T_s6eudgtEKnnMVNdHM7fVPlgrxf/view?usp=sharing](https://drive.google.com/file/d/1GOD5T_s6eudgtEKnnMVNdHM7fVPlgrxf/view?usp=sharing). The places where I have left `"?"` is where I'm having trouble understanding how the gradient will be computed.

**Comment by hello-fri-end** [comment]
Found something that might be helpful for us : [https://github.com/ddbourgin/numpy-ml/blob/master/numpy_ml/neural_nets/losses/losses.py#L457](https://github.com/ddbourgin/numpy-ml/blob/master/numpy_ml/neural_nets/losses/losses.py#L457). I'm not sure why they have added `eps `  to  `X_interp_norm` ; rest of the calculation is the same as my hand-computed version above.

**Comment by zoq** [comment]
`eps` is introduced to avoid division by zero in the term: `(gradInterp / X_interp_norm)`. But looking at the numpy code you posted I think this is exactly what we are looking for, no need to gradient of the gradient for the whole network. I could put a pseudo snippet together if you think that could be helpful?

**Comment by hello-fri-end** [comment]
@zoq Thanks for looking into this. There's no need for pseudo code, I understood how it should be implemented. I'll update the changes in the Dual Optimizer PR.

**Comment by zoq** [comment]
Thanks to you for diving into the issue, that was really helpful.
